File: app/src/main/java/edu/schoolapp/ManageListFragment.java
package edu.schoolapp;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.app.AppCompatActivity;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ListView;
import android.widget.Toast;

import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.firestore.CollectionReference;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.QueryDocumentSnapshot;
import com.google.firebase.firestore.QuerySnapshot;

import java.util.ArrayList;
import java.util.List;

import edu.schoolapp.Model.PrimaryMenuDao;

public class ManageListFragment extends AppCompatActivity {

    private ManageListAdapter manageListAdapter;
    private List<PrimaryMenuModel> manageList;
    private FirebaseFirestore db;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.fragment_manage_list);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        db = FirebaseFirestore.getInstance();

        ListView manageListView = findViewById(R.id.manage_list_view);
        manageListAdapter = new ManageListAdapter(manageList);
        manageListView.setAdapter(manageListAdapter);

        findViewById(R.id.manage_button).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                CollectionReference menulist = db.collection("users").document(SchoolApp.getToken()).collection("primaryMenus");
                if (manageList != null) {
                    for (PrimaryMenuModel menuModel : manageList) {
                        PrimaryMenuDao dao = getPrimaryMenuDao(menuModel);
                        menulist.document(dao.getType().toLowerCase()).set(dao);
                    }
                }
                Toast.makeText(ManageListFragment.this, "Your list update successfully", Toast.LENGTH_SHORT).show();
                finish();
            }
        });

        manageListView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                PrimaryMenuModel manageListModel = manageList.get(position);
                manageListModel.setSelected(!manageListModel.isSelected());
                manageListAdapter.notifyDataSetChanged();
            }
        });

        getPrimaryManageList();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            finish();
        }
        return true;
    }

//    private void addManageList() {
//        manageList.add(new PrimaryMenuModel(R.drawable.todo, "Todo", PrimaryMenuType.TODO));
//        manageList.add(new PrimaryMenuModel(R.drawable.assignment, "Assignment", PrimaryMenuType.TODO));
//        manageList.add(new PrimaryMenuModel(R.drawable.web_link, "Web Link", PrimaryMenuType.TODO));
//        manageList.add(new PrimaryMenuModel(R.drawable.mail, "Mail", PrimaryMenuType.TODO));
//        manageList.add(new PrimaryMenuModel(R.drawable.drive, "Drive", PrimaryMenuType.TODO));
//        manageList.add(new PrimaryMenuModel(R.drawable.schedule, "TimeTable", PrimaryMenuType.TODO));
//        manageList.add(new PrimaryMenuModel(R.drawable.time_table, "Calendar", PrimaryMenuType.TODO));
//    }

    private void getPrimaryManageList() {
        db.collection("users").document(SchoolApp.getToken()).collection("primaryMenus").get().addOnSuccessListener(new OnSuccessListener<QuerySnapshot>() {
            @Override
            public void onSuccess(QuerySnapshot queryDocumentSnapshots) {
                manageList = new ArrayList<>();
                if (!queryDocumentSnapshots.isEmpty()) {
                    for (QueryDocumentSnapshot snapshot : queryDocumentSnapshots) {
                        PrimaryMenuDao pMenu = snapshot.toObject(PrimaryMenuDao.class);
                        manageList.add(getPrimaryMenuModel(pMenu));
                    }
                    manageListAdapter.updateData(manageList);
                }
            }
        });
    }

    private PrimaryMenuModel getPrimaryMenuModel(PrimaryMenuDao dao) {
        PrimaryMenuModel primaryMenuModel = null;
        switch (dao.getType()) {
            case "TODO":
                primaryMenuModel = new PrimaryMenuModel(R.drawable.todo, PrimaryMenuType.TODO.toString(), PrimaryMenuType.TODO, dao.isEnable());
                break;
            case "ASSIGNMENT":
                primaryMenuModel = new PrimaryMenuModel(R.drawable.assignment, PrimaryMenuType.ASSIGNMENT.toString(), PrimaryMenuType.ASSIGNMENT, dao.isEnable());
                break;
            case "WEBLINK":
                primaryMenuModel = new PrimaryMenuModel(R.drawable.web_link, PrimaryMenuType.WEBLINK.toString(), PrimaryMenuType.WEBLINK, dao.isEnable());
                break;
            case "MAIL":
                primaryMenuModel = new PrimaryMenuModel(R.drawable.mail, PrimaryMenuType.MAIL.toString(), PrimaryMenuType.MAIL, dao.isEnable());
                break;
            case "DRIVE":
                primaryMenuModel = new PrimaryMenuModel(R.drawable.drive, PrimaryMenuType.DRIVE.toString(), PrimaryMenuType.DRIVE, dao.isEnable());
                break;
            case "TIMETABLE":
                primaryMenuModel = new PrimaryMenuModel(R.drawable.schedule, PrimaryMenuType.TIMETABLE.toString(), PrimaryMenuType.TIMETABLE, dao.isEnable());
                break;
            case "CALENDAR":
                primaryMenuModel = new PrimaryMenuModel(R.drawable.time_table, PrimaryMenuType.CALENDAR.toString(), PrimaryMenuType.CALENDAR, dao.isEnable());
                break;
        }
        return primaryMenuModel;
    }

    private PrimaryMenuDao getPrimaryMenuDao(PrimaryMenuModel menuModel) {
        PrimaryMenuDao menuDao = null;
        switch (menuModel.getMenuType()) {
            case TODO:
                menuDao = new PrimaryMenuDao(menuModel.getMenuName(), "TODO", menuModel.isSelected());
                break;
            case ASSIGNMENT:
                menuDao = new PrimaryMenuDao(menuModel.getMenuName(), "ASSIGNMENT", menuModel.isSelected());
                break;
            case MAIL:
                menuDao = new PrimaryMenuDao(menuModel.getMenuName(), "MAIL", menuModel.isSelected());
                break;
            case DRIVE:
                menuDao = new PrimaryMenuDao(menuModel.getMenuName(), "DRIVE", menuModel.isSelected());
                break;
            case WEBLINK:
                menuDao = new PrimaryMenuDao(menuModel.getMenuName(), "WEBLINK", menuModel.isSelected());
                break;
            case TIMETABLE:
                menuDao = new PrimaryMenuDao(menuModel.getMenuName(), "TIMETABLE", menuModel.isSelected());
                break;
            case CALENDAR:
                menuDao = new PrimaryMenuDao(menuModel.getMenuName(), "CALENDAR", menuModel.isSelected());
                break;
        }
        return menuDao;
    }
}

File: app/src/main/java/edu/schoolapp/ManageListFragment.kt
package edu.schoolapp

import android.os.Bundle
import android.view.MenuItem
import android.view.View
import android.widget.AdapterView.OnItemClickListener
import android.widget.ListView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.QueryDocumentSnapshot
import edu.schoolapp.model.PrimaryMenuDao
import edu.schoolapp.SchoolApp.Companion.token
import java.util.Locale

class ManageListFragment : AppCompatActivity() {
    private var manageListAdapter: ManageListAdapter? = null
    private var manageList: MutableList<PrimaryMenuModel>? = null
    private lateinit var db: FirebaseFirestore
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.fragment_manage_list)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        db = FirebaseFirestore.getInstance()
        val manageListView = findViewById<ListView>(R.id.manage_list_view)
        manageListAdapter = ManageListAdapter(manageList)
        manageListView.adapter = manageListAdapter
        findViewById<View>(R.id.manage_button).setOnClickListener {
            val menuList = db.collection("users").document((token)!!).collection("primaryMenus")
            if (manageList != null) {
                for (menuModel: PrimaryMenuModel? in manageList!!) {
                    val dao = getPrimaryMenuDao(menuModel)
                    if (dao != null) {
                        menuList.document(dao.type!!.lowercase(Locale.getDefault())).set(dao)
                    }
                }
            }
            Toast.makeText(
                this@ManageListFragment,
                "Your list update successfully",
                Toast.LENGTH_SHORT
            ).show()
            finish()
        }
        manageListView.onItemClickListener = OnItemClickListener { _, _, position, _ ->
                val manageListModel: PrimaryMenuModel = manageList!![position]
                manageListModel.isSelected = !manageListModel.isSelected
                manageListAdapter!!.notifyDataSetChanged()
            }

        getPrimaryManageList()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            finish()
        }
        return true
    }

    private fun getPrimaryManageList() {
        db.collection("users").document((token)!!).collection("primaryMenus").get()
            .addOnSuccessListener { queryDocumentSnapshots ->
                manageList = ArrayList()
                if (!queryDocumentSnapshots.isEmpty) {
                    for (snapshot: QueryDocumentSnapshot in queryDocumentSnapshots) {
                        val pMenu = snapshot.toObject(
                            PrimaryMenuDao::class.java
                        )
                        getPrimaryMenuModel(pMenu)?.let { manageList?.add(it) }
                    }
                    manageListAdapter!!.updateData(manageList)
                }
            }
    }

    private fun getPrimaryMenuModel(dao: PrimaryMenuDao): PrimaryMenuModel? {
        var primaryMenuModel: PrimaryMenuModel? = null
        when (dao.type) {
            "TODO" -> primaryMenuModel = PrimaryMenuModel(
                R.drawable.todo,
                PrimaryMenuType.TODO.toString(),
                PrimaryMenuType.TODO,
                dao.isEnable
            )

            "ASSIGNMENT" -> primaryMenuModel = PrimaryMenuModel(
                R.drawable.assignment,
                PrimaryMenuType.ASSIGNMENT.toString(),
                PrimaryMenuType.ASSIGNMENT,
                dao.isEnable
            )

            "WEBLINK" -> primaryMenuModel = PrimaryMenuModel(
                R.drawable.web_link,
                PrimaryMenuType.WEBLINK.toString(),
                PrimaryMenuType.WEBLINK,
                dao.isEnable
            )

            "MAIL" -> primaryMenuModel = PrimaryMenuModel(
                R.drawable.mail,
                PrimaryMenuType.MAIL.toString(),
                PrimaryMenuType.MAIL,
                dao.isEnable
            )

            "DRIVE" -> primaryMenuModel = PrimaryMenuModel(
                R.drawable.drive,
                PrimaryMenuType.DRIVE.toString(),
                PrimaryMenuType.DRIVE,
                dao.isEnable
            )

            "TIMETABLE" -> primaryMenuModel = PrimaryMenuModel(
                R.drawable.schedule,
                PrimaryMenuType.TIMETABLE.toString(),
                PrimaryMenuType.TIMETABLE,
                dao.isEnable
            )

            "CALENDAR" -> primaryMenuModel = PrimaryMenuModel(
                R.drawable.time_table,
                PrimaryMenuType.CALENDAR.toString(),
                PrimaryMenuType.CALENDAR,
                dao.isEnable
            )
        }
        return primaryMenuModel
    }

    private fun getPrimaryMenuDao(menuModel: PrimaryMenuModel?): PrimaryMenuDao? {
        var menuDao: PrimaryMenuDao? = null
        when (menuModel!!.menuType) {
            PrimaryMenuType.TODO -> menuDao =
                PrimaryMenuDao(menuModel.menuName, "TODO", menuModel.isSelected)

            PrimaryMenuType.ASSIGNMENT -> menuDao =
                PrimaryMenuDao(menuModel.menuName, "ASSIGNMENT", menuModel.isSelected)

            PrimaryMenuType.MAIL -> menuDao =
                PrimaryMenuDao(menuModel.menuName, "MAIL", menuModel.isSelected)

            PrimaryMenuType.DRIVE -> menuDao =
                PrimaryMenuDao(menuModel.menuName, "DRIVE", menuModel.isSelected)

            PrimaryMenuType.WEBLINK -> menuDao =
                PrimaryMenuDao(menuModel.menuName, "WEBLINK", menuModel.isSelected)

            PrimaryMenuType.TIMETABLE -> menuDao =
                PrimaryMenuDao(menuModel.menuName, "TIMETABLE", menuModel.isSelected)

            PrimaryMenuType.CALENDAR -> menuDao =
                PrimaryMenuDao(menuModel.menuName, "CALENDAR", menuModel.isSelected)

            else -> {}
        }
        return menuDao
    }
}